### ISSUE TITLE
XrdHttp: "TE" header + "X-Transfer-Status" + byte-range read results in bad transfer data

### ISSUE THREAD

**Issue by matyasselmeci** [xrootd/xrootd#2076]
On XRootD 5.6.1 with the HTTP plugin, the following combination of HTTP headers causes bad data to be returned:
```
X-Transfer-Status: true
TE: trailers
Range: bytes 0-
```
(any byte range will work, not just `0-`)

namely instead of returning the file contents, it returns the content length in hex, followed by a CR/LF, followed by the actual data (minus however many bytes the length and CR/LF took).

This curl command can reproduce it (assuming a server on localhost with XrdHttp listening on 1094 and a file named `/testfile`):
```
curl http://localhost:1094//testfile -i -H "Range: bytes 0-" -H "X-Transfer-Status: true" -H "TE: trailers"
```

**Comment by abh3** [comment]
Yes, it would appear that it does not properly handle  a start only range, which is allowed.
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range

**Comment by smithdh** [comment]
Hi. I'm working on a fix for #1976; and this problem appears to be in a similar area. So I proposed to take this also, Cedric agreed, so I'l change over the assignment.

**Comment by smithdh** [comment]
Thanks @bbockelm @matyasselmeci 

The fix which Brian described (which I interpreted as snippet (x) below) is needed; it's quite local and doesn't really overlap with #1976; with a fully specified range like 100-200 it solves the current issue. But during tests I've seen it can leave curl hanging (waiting for the last ('0') chunk), when given an open range like '100-' (but not '0-'). This comes about because we have some problems handling those ranges. That should be fixed, which seems to go overlap a lot with what I'm preparing for #1976.

So if it's ok I'd hold off with a PR for this until I get a fix for #1976 ready (but that might be another ~week), and then put a fix for this with that work (or make the PR at the same time).

--
(x) tried with this change:

{code}
diff --git a/src/XrdHttp/XrdHttpReq.cc b/src/XrdHttp/XrdHttpReq.cc
index e119181..d601f47 100644
--- a/src/XrdHttp/XrdHttpReq.cc
+++ b/src/XrdHttp/XrdHttpReq.cc
@@ -2214,7 +2214,11 @@ int XrdHttpReq::PostProcessHTTPReq(bool final_) {
                   s += responseHeader.c_str();
                 }
 
-                prot->SendSimpleResp(206, NULL, (char *)s.c_str(), NULL, cnt, keepalive);
+                if (m_transfer_encoding_chunked && m_trailer_headers) {
+                  prot->StartChunkedResp(206, NULL, (char *)s.c_str(), cnt, keepalive);
+                } else {
+                  prot->SendSimpleResp(206, NULL, (char *)s.c_str(), NULL, cnt, keepalive);
+                }
                 return 0;
               } else
                 if (rwOps.size() > 1) {
{code}

**Comment by bbockelm** [comment]
@amadio - does David's suggestion above jive with your plans for 5.6.2?

**Comment by amadio** [comment]
Yes, we discussed this today. The plan is to have 5.6.2 ready by the end of next week more or less. David's patch should be ready by then, otherwise we release 5.6.2 and as soon as David is done we release 5.6.3 to include the fix for this (or delay 5.6.2 depending on the circumstances).

Also, note that in the `devel` branch we have already a fix for a related issue with chunked tranfers, so it's important to develop any fix on top of `devel`.

**Comment by smithdh** [comment]
hi @bbockelm I noticed a couple of things;

We have problems with multiple ranges and chunked encoding (i.e. "multipart/byteranges"). This is very similar to the current issue, shall I fix that too?

I noticed we return a Content-Length header also when using chunked transfer encoding. I don't think the server should do that (rfc2616 section 4.4, point 3). I was wondering if you have an opinion as to whether we should suppress this header in chunked transfer-encoding?

**Comment by bbockelm** [comment]
> This is very similar to the current issue, shall I fix that too?

Yes - please do!  We were starting with the minimal reproducer, we haven't gotten much beyond that.

(N.b. -- should we be starting to add a few simple unit tests around these items?  Since we've got an independent implementation of the protocol - `libcurl` - it should be relatively straightforward)

> I was wondering if you have an opinion as to whether we should suppress this header in chunked transfer-encoding?

No opinion -- your interpretation seems correct.

**Comment by amadio** [comment]
> should we be starting to add a few simple unit tests around these items? S

Yes, I plan to add many tests on all areas we currently do not test, like http (some tests exist, but need to be converted and plugged to CMake), and auth methods, starting with OpenSSL and Krb5, but later also tokens. Our Summer student make great progress towards that which will be merged soon, and I will continue.

**Comment by amadio** [comment]
Fixed by #2072.
